### PR TITLE
Optimize the is_permutation family and _Hash::operator== for multicontainers

### PR DESCRIPTION
**Pull request by BillyONeal**
`<xutility>`  
4660: _Find_pr is a helper for is_permutation, so move it down to that area.  
4684: The SHOUTY banners were attached to functions which were implementation details of is_permutation, so I fixed them up to say is_permutation and removed the banners for helper functions.  
4711: Use if constexpr to avoid a tag dispatch call for _Trim_matching_suffixes. Optimizers will like this because they generally hate reference-to-pointer, and it also serves to workaround DevCom-883631 when this algorithm is constexprized.  
4766: Indicate that we are trimming matching prefixes in this loop body, and break apart comment block that was incorrectly merged by clang-format.  
4817: In the dual range forward version of the algorithm, calculate the distances concurrently to avoid wasting lots of time when the distances vary by a lot. For example, is_permutation( a forward range of length 1, a forward range of length 1'000'000 ) used to do the million increments, now it stops at 1 increment.  
4862: In the dual range random-access version, avoid recalculating _Last2 when it has already been supplied to us.  
  
`<xhash>`  
1404: Move down construction of _Bucket_hi in _Equal_range to before the first loop body using it.  
1918: Added a new function to calculate equality for unordered multicontainers. We loop over the elements in the left container, find corresponding ranges in the right container, trim prefixes, then dispatch to is_permutation's helper _Check_match_counts.  
Improvements over the old implementation:  
* For standard containers, we no longer need to hash any elements from the left container; we know that we've found the "run" of equivalent elements because we *started* with an element in that container. We also never go "backwards" or multiply enumerate _Left (even for !_Standard), which improves cache use when the container becomes large.
* Just like the dual range is_permutation improvement above, when the equal_ranges of the containers are of wildly varying lengths, this will stop on the shorter of the lengths.
* We avoid the 3-arg is_permutation doing a linear time operation to discover _Last2 that we already had calculated in determining _Right's equal_range.

The function _Multi_equal_check_equal_range tests one equal_range from the left container against the corresponding equal_range from the right container, while _Multi_equal invokes _Multi_equal_check_equal_range for each equal_range.

Performance results:

image

Benchmark code:
```
#undef NDEBUG
#define _SILENCE_STDEXT_HASH_DEPRECATION_WARNINGS
#include <assert.h>
#include <benchmark/benchmark.h>
#include <hash_map>
#include <random>
#include <stddef.h>
#include <unordered_map>
#include <utility>
#include <vector>

using namespace std;

template <template <class...> class MapType> void HashRandomUnequal(benchmark::State &state) {
    std::minstd_rand rng(std::random_device{}());
    const auto range0 = static_cast<ptrdiff_t>(state.range(0));
    vector<pair<unsigned, unsigned>> testData;
    testData.resize(range0 * 5);
    const auto dataEnd = testData.begin() + range0;
    std::generate(testData.begin(), dataEnd, [&]() { return pair<unsigned, unsigned>{rng(), 0u}; });
    std::copy(testData.begin(), dataEnd,
              std::copy(testData.begin(), dataEnd,
                        std::copy(testData.begin(), dataEnd, std::copy(testData.begin(), dataEnd, dataEnd))));
    std::unordered_multimap<unsigned, unsigned> a(testData.begin(), testData.end());
    testData.clear();
    std::unordered_multimap<unsigned, unsigned> b = a;
    next(b.begin(), b.size() - 1)->second = 1u;
    for (auto &&_ : state) {
        (void)_;
        assert(a != b);
    }
}

BENCHMARK_TEMPLATE1(HashRandomUnequal, unordered_multimap)->Arg(1)->Arg(10)->Range(100, 100'000);
BENCHMARK_TEMPLATE1(HashRandomUnequal, hash_multimap)->Arg(1)->Arg(10)->Range(100, 100'000);

template <template <class...> class MapType> void HashRandomEqual(benchmark::State &state) {
    std::minstd_rand rng(std::random_device{}());
    const auto range0 = static_cast<ptrdiff_t>(state.range(0));
    vector<pair<unsigned, unsigned>> testData;
    testData.resize(range0 * 5);
    const auto dataEnd = testData.begin() + range0;
    std::generate(testData.begin(), dataEnd, [&]() { return pair<unsigned, unsigned>{rng(), 0}; });
    std::copy(testData.begin(), dataEnd,
              std::copy(testData.begin(), dataEnd,
                        std::copy(testData.begin(), dataEnd, std::copy(testData.begin(), dataEnd, dataEnd))));
    std::unordered_multimap<unsigned, unsigned> a(testData.begin(), testData.end());
    testData.clear();
    std::unordered_multimap<unsigned, unsigned> b = a;
    for (auto &&_ : state) {
        (void)_;
        assert(a == b);
    }
}

BENCHMARK_TEMPLATE1(HashRandomEqual, unordered_multimap)->Arg(1)->Arg(10)->Range(100, 100'000);
BENCHMARK_TEMPLATE1(HashRandomEqual, hash_multimap)->Arg(1)->Arg(10)->Range(100, 100'000);

template <template <class...> class MapType> void HashUnequalDifferingBuckets(benchmark::State &state) {
    std::unordered_multimap<unsigned, unsigned> a;
    std::unordered_multimap<unsigned, unsigned> b;
    const auto range0 = static_cast<ptrdiff_t>(state.range(0));
    for (ptrdiff_t idx = 0; idx < range0; ++idx) {
        a.emplace(0, 1);
        b.emplace(1, 0);
    }

    a.emplace(1, 0);
    b.emplace(0, 1);
    for (auto &&_ : state) {
        (void)_;
        assert(a != b);
    }
}

BENCHMARK_TEMPLATE1(HashUnequalDifferingBuckets, unordered_multimap)->Arg(2)->Arg(10)->Range(100, 100'000);
BENCHMARK_TEMPLATE1(HashUnequalDifferingBuckets, hash_multimap)->Arg(2)->Arg(10)->Range(100, 100'000);

BENCHMARK_MAIN();
```

# Checklist

Be sure you've read README.md and understand the scope of this repo.

If you're unsure about a box, leave it unchecked. A maintainer will help you.

- [x] Identifiers in product code changes are properly `_Ugly` as per
  https://eel.is/c++draft/lex.name#3.1 or there are no product code changes.
- [x] The STL builds successfully and all tests have passed (must be manually
  verified by an STL maintainer before automated testing is enabled on GitHub,
  leave this unchecked for initial submission).
- [x] These changes introduce no known ABI breaks (adding members, renaming
  members, adding virtual functions, changing whether a type is an aggregate
  or trivially copyable, etc.).
- [x] These changes were written from scratch using only this repository,
  the C++ Working Draft (including any cited standards), other WG21 papers
  (excluding reference implementations outside of proposed standard wording),
  and LWG issues as reference material. If they were derived from a project
  that's already listed in NOTICE.txt, that's fine, but please mention it.
  If they were derived from any other project (including Boost and libc++,
  which are not yet listed in NOTICE.txt), you *must* mention it here,
  so we can determine whether the license is compatible and what else needs
  to be done.

### CODE REVIEW

**Comment by BillyONeal** [comment]
I am specifically interested in more perf tests folks think I should run.

**Comment by CaseyCarter** [comment]
> The SHOUTY banners were attached to functions which were implmentation details

s/implmentation/implementation/

**Comment by BillyONeal** [comment]
> > The SHOUTY banners were attached to functions which were implmentation details
> 
> s/implmentation/implementation/

Fixed, feel free to just edit stuff like that :P

**Comment by CaseyCarter** [comment]
> Fixed, feel free to just edit stuff like that :P

I started to, but was concerned you might have a dual internal PR that needs to be corrected also.

**Comment by BillyONeal** [comment]
> > Fixed, feel free to just edit stuff like that :P
> 
> I started to, but was concerned you might have a dual internal PR that needs to be corrected also.

There are no test changes needed for this so I'm going to let it be approved here before making a dual one

File: stl/inc/xhash [hunk 1]
```diff
@@ -1915,6 +1915,169 @@ protected:
...
+        const _Hash& _Right, _Unchecked_const_iterator _First1) const {
+        // check equal_range of elements matching *_First1 match the equal_range of elements in _Right
+        auto& _Keyval = _Traits::_Kfn(*_First1);
+        // find the start of the matching run in the other container
+        const size_t _Hashval   = _Right._Traitsobj(_Keyval);
+        const size_type _Bucket = _Hashval & _Right._Mask;
+        auto _First2            = _Right._Vec._Mypair._Myval2._Myfirst[_Bucket << 1];
+        if (_First2 == _Right._Unchecked_end()) {
+            // no matching bucket, therefore no matching run
+            return {false, {}};
```
Comment (CaseyCarter) [inline stl/inc/xhash]: `return {};` would work here. (Several occurrences)

It's arguable that the longer version communicates more information to the reader, but I think `return {};` is actually a clearer indication that there's no useful result.

File: stl/inc/xhash [hunk 2]
```diff
@@ -1915,6 +1915,169 @@ protected:
...
+                    }
+
+                    if (_Left_range_end || _Right_range_end) {
+                        // different number of elements in the range, fail
+                        return {false, {}};
+                    }
+                }
+            }
+        else {
+            // equivalence check the first element when !_Standard
```
Comment (CaseyCarter) [inline stl/inc/xhash]: I found this comment hard to parse:
```suggestion
            // check the first elements for equivalence when !_Standard
```

File: stl/inc/xhash [hunk 3]
```diff
@@ -1915,6 +1915,169 @@ protected:
...
+                    const bool _Left_range_end =
+                        _First1 == _Unchecked_end() || _Traitsobj(_Keyval, _Traits::_Kfn(*_First1));
+
+                    if (_Left_range_end && _Right_range_end) {
+                        // the equal_ranges were completely equal
+                        return {true, _First1};
+                    }
+
+                    if (_Left_range_end || _Right_range_end) {
+                        // one equal_range is a prefix of the other; different lengths mean not equal though
```
Comment (CaseyCarter) [inline stl/inc/xhash]: Missing a "," before "though", but honestly I think this could simply be:
```suggestion
                        // one equal_range is a prefix of the other; not equal
```

File: stl/inc/xhash [hunk 4]
```diff
@@ -1915,6 +1915,169 @@ protected:
...
+                }
+            }
+        }
+    }
+
+    _NODISCARD bool _Multi_equal(const _Hash& _Right) const {
+        static_assert(_Traits::_Multi, "This function only works with multi containers");
+        _STL_INTERNAL_CHECK(this->size() == _Right.size());
+
+        const auto _End = _Right._Unchecked_end();
```
Comment (CaseyCarter) [inline stl/inc/xhash]: Unused.

File: stl/inc/xutility [hunk 5]
```diff
@@ -4750,15 +4760,15 @@ bool _Check_match_counts(_FwdIt1 _First1, _FwdIt1 _Last1, _FwdIt2 _First2, _FwdI
...
 }
 
-// FUNCTION TEMPLATE is_permutation
 template <class _FwdIt1, class _FwdIt2, class _Pr>
-bool _Is_permutation_unchecked(_FwdIt1 _First1, _FwdIt1 _Last1, _FwdIt2 _First2, _Pr _Pred) {
+_NODISCARD bool _Is_permutation_unchecked(_FwdIt1 _First1, _FwdIt1 _Last1, _FwdIt2 _First2, _Pr _Pred) {
     // test if [_First1, _Last1) == permuted [_First2, ...), using _Pred
-    for (; _First1 != _Last1; ++_First1, (void) ++_First2) {
+    for (; _First1 != _Last1; ++_First1, (void) ++_First2) { // trim matching prefix
         if (!_Pred(*_First1, *_First2)) {
```
[the remark this answers is not in the document]
Comment (BillyONeal) [reply stl/inc/xutility]: Option 3: I think we should fix them, but I think they should be fixed in an isolated commit that adds testing for it. (Tweaking our example functors that are passed through the "instantiate all algorithms with all iterator powers" test should be relatively easy)

File: stl/inc/xutility [hunk 6]
```diff
@@ -4804,17 +4814,41 @@ template <class _FwdIt1, class _FwdIt2, class _Pr>
 bool _Is_permutation_unchecked(_FwdIt1 _First1, _FwdIt1 _Last1, _FwdIt2 _First2, _FwdIt2 _Last2, _Pr _Pred,
     forward_iterator_tag, forward_iterator_tag) {
     // test if [_First1, _Last1) == permuted [_First2, _Last2), using _Pred, arbitrary iterators
-    for (; _First1 != _Last1 && _First2 != _Last2; ++_First1, (void) ++_First2) {
+    for (;;) { // trim matching prefix
+        if (_First1 == _Last1) {
+            return _First2 == _Last2; // sequence equal
+        }
+
+        if (_First2 == _Last2) {
```
[the remark this answers is not in the document]
Comment (BillyONeal) [reply stl/inc/xutility]: Codegen for that appears worse: https://gcc.godbolt.org/z/oHGQ6r

[the remark this answers is not in the document]
Comment (BillyONeal) [reply stl/inc/xutility]: image

File: stl/inc/xhash (hunk 1 above)
Comment (BillyONeal) [reply stl/inc/xhash]: I mean, the false is a useful result but I'm OK with shorter here.

File: stl/inc/xutility (hunk 6 above)
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/xutility]: I've proven once again that I can't reason about MSVC's optimizer.

File: stl/inc/xhash (hunk 1 above)
Comment (CaseyCarter) [reply stl/inc/xhash]: Yeah, "useful" is the wrong word.

File: stl/inc/xutility (hunk 6 above)
[the remark this answers is not in the document]
Comment (BillyONeal) [reply stl/inc/xutility]: FWIW LLVM says that those are identical: 
image

[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/xutility]: Yes, so does GCC. I was expecting identical codegen for pointers, and better for less-analyzable iterators. 🤷‍♂

[the remark this answers is not in the document]
Comment (miscco) [reply stl/inc/xutility]: That sounds like a worthy "please improve codegen for common usecase" ticket

[the remark this answers is not in the document]
Comment (BillyONeal) [reply stl/inc/xutility]: > That sounds like a worthy "please improve codegen for common usecase" ticket

I won't file bugs like that without a performance test showing the difference and I think it would be very difficult to meaningfully measure the difference here.

File: stl/inc/xhash [hunk 7]
```diff
@@ -1915,6 +1915,166 @@ protected:
...
+        else {
+            // check the first elements for equivalence when !_Standard
+            if (_Right._Traitsobj(_Keyval, _Traits::_Kfn(*_First2))) {
+                return {};
+            }
+
+            // trim matching prefixes
+            const size_t _LHashval   = _Traitsobj(_Keyval);
+            const size_type _LBucket = _LHashval & _Mask;
+            const auto _LBucket_hi   = _Vec._Mypair._Myval2._Myfirst[(_LBucket << 1) + 1];
```
Comment (miscco) [inline stl/inc/xhash]: If I had sufficient coffee this time this is the main difference between the two `if constexpr` branches.

I you would store the pointer after `_LBucket_hi` you could also write

```cpp
    ++_First1;
    const bool _Left_range_end =  
        _First1 == _LBucket_hi_next || _Traitsobj(_Keyval, _Traits::_Kfn(*_First1));
```

With that the only difference between the two loops (besides comments) is `_LBucket_hi_next ` vs `_Unchecked_end()`. Given that you use a temporary variable anyway you could assign `_Unchecked_end()` to `_LBucket_hi_next ` in the `_Traits::_Standard` case

Dunno if that is truly worth it

Comment (BillyONeal) [reply stl/inc/xhash]: That will do one extra pointer dereference per bucket (to find `_LBucket_hi_next`); but maybe we don't care about the perf of `!_Standard` (that is, `hash_foo` rather than `unordered_foo`)  ...

File: stl/inc/xutility [hunk 8]
```diff
@@ -4693,10 +4681,20 @@ _NODISCARD _Iter_diff_t<_FwdIt> count(
     _ExPo&& _Exec, const _FwdIt _First, const _FwdIt _Last, const _Ty& _Val) noexcept; // terminates
 #endif // _HAS_CXX17
 
-// FUNCTION TEMPLATE _Count_pr
+// FUNCTION TEMPLATE is_permutation
 template <class _InIt, class _Ty, class _Pr>
-_Iter_diff_t<_InIt> _Count_pr(_InIt _First, _InIt _Last, const _Ty& _Val, _Pr _Pred) {
-    // count elements that match _Val, using _Pred
+_NODISCARD _InIt _Find_pr(_InIt _First, const _InIt _Last, const _Ty& _Val, _Pr _Pred) {
```
Comment (miscco) [inline stl/inc/xutility]: Could that be `constexpr` given that it essentially is a derivate of `std::find_if`

Comment (BillyONeal) [reply stl/inc/xutility]: Yes it can, and indeed #425 makes that change :)

File: stl/inc/xhash (hunk 3 above)
Comment (CaseyCarter) [reply stl/inc/xhash]: or even `// one equal_range is a proper prefix of the other`

File: stl/inc/xutility [hunk 9]
```diff
@@ -4726,12 +4724,24 @@ void _Trim_matching_suffixes(
...
-// FUNCTION TEMPLATE _Check_match_counts
 template <class _FwdIt1, class _FwdIt2, class _Pr>
-bool _Check_match_counts(_FwdIt1 _First1, _FwdIt1 _Last1, _FwdIt2 _First2, _FwdIt2 _Last2, _Pr _Pred) {
-    // test if [_First1, _Last1) == permuted [_First2, _Last2), using _Pred, same lengths
+_NODISCARD bool _Check_match_counts(_FwdIt1 _First1, _FwdIt1 _Last1, _FwdIt2 _First2, _FwdIt2 _Last2, _Pr _Pred) {
+    // test if [_First1, _Last1) == permuted [_First2, _Last2), after matching prefix removal
+    _STL_INTERNAL_CHECK(_STD distance(_First1, _Last1) == _STD distance(_First2, _Last2));
+#if _HAS_IF_CONSTEXPR
+    if constexpr (_Is_bidi_iter_v<_FwdIt1> && _Is_bidi_iter_v<_FwdIt2>) {
+        do { // find last inequality
```
[the remark this answers is not in the document]
Comment (BillyONeal) [reply stl/inc/xutility]: This is now documented by the first `_STL_INTERNAL_CHECK`.

[the remark this answers is not in the document]
Comment (BillyONeal) [reply stl/inc/xutility]: Hmmm I'm wondering if I dropped changes on the floor because I'm not seeing it... damnit, sorry.

[the remark this answers is not in the document]
Comment (BillyONeal) [reply stl/inc/xutility]: Ah, the change *is* here, but when I follow the link from your comment it goes to the old commit :)

**Comment by BillyONeal** [comment]
> My non-resolved comments are suggestions, so I "Approve with suggestions."

Is there anything I missed other than the `!_Pred` thing?

**Comment by CaseyCarter** [comment]
> Is there anything I missed other than the `!_Pred` thing?

No - I think I was confused by clicking the through the comments and observing old state which I thought was current state.

File: stl/inc/xhash (hunk 7 above)
Comment (BillyONeal) [reply stl/inc/xhash]: Looks good, applied.

File: stl/inc/xutility [hunk 10]
```diff
@@ -4693,10 +4681,20 @@ _NODISCARD _Iter_diff_t<_FwdIt> count(
     _ExPo&& _Exec, const _FwdIt _First, const _FwdIt _Last, const _Ty& _Val) noexcept; // terminates
 #endif // _HAS_CXX17
 
-// FUNCTION TEMPLATE _Count_pr
+// FUNCTION TEMPLATE is_permutation
```
Comment (StephanTLavavej) [inline stl/inc/xutility]: FYI, #279 tracks a possible improvement to `is_permutation`, although it isn't necessary to implement at this time.

File: stl/inc/xhash [hunk 11]
```diff
@@ -1915,6 +1915,116 @@ protected:
         return _List._Getal();
     }
 
+    struct _Multi_equal_check_result {
+        bool _Equal_possible;
```
Comment (StephanTLavavej) [inline stl/inc/xhash]: I am mildly uncomfortable with the lack of DMIs here; `return {}` initializes this `bool` to `false` but the risk of garbage-init is higher than I would like. Would it be better to add DMIs for `{false}` and `{}` for the iterator?

File: stl/inc/xhash [hunk 12]
```diff
@@ -1915,6 +1915,116 @@ protected:
...
     }
 
+    struct _Multi_equal_check_result {
+        bool _Equal_possible;
+        _Unchecked_const_iterator _Subsequent_first; // only useful if _Equal_possible
+    };
+
+    _NODISCARD _Multi_equal_check_result _Multi_equal_check_equal_range(
+        const _Hash& _Right, _Unchecked_const_iterator _First1) const {
+        // check equal_range of elements matching *_First1 match the equal_range of elements in _Right
```
Comment (StephanTLavavej) [inline stl/inc/xhash]: This comment grammar is difficult to parse.

File: stl/inc/xhash [hunk 13]
```diff
@@ -1915,6 +1915,116 @@ protected:
...
+            }
+        }
+
+        _Unchecked_const_iterator _Left_stop_at;
+        if
+            _CONSTEXPR_IF(_Traits::_Standard) {
+                _Left_stop_at = _Unchecked_end();
+            }
+        else {
+            // check the first elements for equivalence when !_Standard
```
Comment (StephanTLavavej) [inline stl/inc/xhash]: Do we have sufficient correctness tests for both `unordered_meow` and `hash_meow`? You're introducing new logic for `!_Standard` so the risk of breaking `hash_meow` is elevated.

File: stl/inc/xhash [hunk 14]
```diff
@@ -1915,6 +1915,116 @@ protected:
...
+        }
+    }
+
+    _NODISCARD bool _Multi_equal(const _Hash& _Right) const {
+        static_assert(_Traits::_Multi, "This function only works with multi containers");
+        _STL_INTERNAL_CHECK(this->size() == _Right.size());
+        const auto _Last1 = _Unchecked_end();
+        auto _First1      = _Unchecked_begin();
+        while (_First1 != _Last1) {
+            auto _Result = _Multi_equal_check_equal_range(_Right, _First1);
```
Comment (StephanTLavavej) [inline stl/inc/xhash]: Can `_Result` be `const`?

File: stl/inc/xutility [hunk 15]
```diff
@@ -4726,39 +4724,53 @@ void _Trim_matching_suffixes(
...
-    for (; _First1 != _Last1; ++_First1, (void) ++_First2) {
+    for (; _First1 != _Last1; ++_First1, (void) ++_First2) { // trim matching prefix
         if (!_Pred(*_First1, *_First2)) {
-            // found first inequality, check match counts in suffix narrowing _Iter_diff_t<_FwdIt1> to
-            // _Iter_diff_t<_FwdIt2> is OK because if the 2nd range is shorter than the 1st, the user already
-            // triggered UB
+            // found first inequality, check match counts in suffix
+            //
+            // narrowing _Iter_diff_t<_FwdIt1> to _Iter_diff_t<_FwdIt2> is OK because if the 2nd range is shorter than
+            // the 1st, the user already triggered UB
```
Comment (StephanTLavavej) [inline stl/inc/xutility]: Style: I recommend saying "second" and "first" instead of aggressively abbreviating.

File: stl/inc/xutility [hunk 16]
```diff
@@ -4804,17 +4816,41 @@ template <class _FwdIt1, class _FwdIt2, class _Pr>
 bool _Is_permutation_unchecked(_FwdIt1 _First1, _FwdIt1 _Last1, _FwdIt2 _First2, _FwdIt2 _Last2, _Pr _Pred,
     forward_iterator_tag, forward_iterator_tag) {
     // test if [_First1, _Last1) == permuted [_First2, _Last2), using _Pred, arbitrary iterators
-    for (; _First1 != _Last1 && _First2 != _Last2; ++_First1, (void) ++_First2) {
+    for (;;) { // trim matching prefix
+        if (_First1 == _Last1) {
+            return _First2 == _Last2; // sequences are equal
```
Comment (StephanTLavavej) [inline stl/inc/xutility]: This comment is only true if we return true, although I suppose it is fairly obvious.

File: stl/inc/xhash (hunk 11 above)
Comment (BillyONeal) [reply stl/inc/xhash]: In my tests all the functions returning this type were inlined so I have no objection to DMIs.

File: stl/inc/xhash (hunk 13 above)
Comment (BillyONeal) [reply stl/inc/xhash]: Yes, the devcrt `Dev11_*****_container_equality` test has a bunch of edge cases tested here.

File: stl/inc/xutility (hunk 15 above)
Comment (BillyONeal) [reply stl/inc/xutility]: I agree, though technically I didn't edit that here :P. Will fix!

File: stl/inc/xutility (hunk 10 above)
Comment (BillyONeal) [reply stl/inc/xutility]: I do agree that looks interesting but would prefer that be a separate change (as it'll need its own perf tests and similar justification)
